File: bindings/zig/src/interfaces.zig
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

const c = @import("c.zig");
const types = @import("types.zig");
const util = @import("util.zig");

pub fn list(node: types.Node, out: []types.InterfaceEntry) types.Error!usize {
    if (out.len == 0) return error.InvalidArg;
    var written: usize = 0;
    const code = c.rns_interfaces(types.asU64(node), out.ptr, out.len, &written);
    if (code != c.RNS_OK and code != c.RNS_ERR_TRUNCATED) {
        try types.mapCode(code);
    }
    return written;
}

pub fn name(entry: *const types.InterfaceEntry) []const u8 {
    return util.cstringField(&entry.name);
}

pub fn typeName(entry: *const types.InterfaceEntry) []const u8 {
    return util.cstringField(&entry.type_name);
}

File: bindings/rust/src/interfaces.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

use crate::error::{map_code, Error, Result};
use crate::ffi::{self, InterfaceEntry};
use crate::node::Node;

#[derive(Debug, Clone)]
pub struct InterfaceInfo {
    pub name: String,
    pub type_name: String,
    pub online: bool,
    pub enabled: bool,
    pub rx_bytes: u64,
    pub tx_bytes: u64,
    pub rx_packets: u64,
    pub tx_packets: u64,
}

fn cstr_field(buf: &[i8]) -> String {
    let bytes: Vec<u8> = buf
        .iter()
        .map(|b| *b as u8)
        .take_while(|b| *b != 0)
        .collect();
    String::from_utf8_lossy(&bytes).into_owned()
}

pub fn interfaces_list(node: &Node, capacity: usize) -> Result<Vec<InterfaceInfo>> {
    if capacity == 0 {
        return Err(Error::InvalidArg);
    }
    let mut raw = vec![unsafe { std::mem::zeroed::<InterfaceEntry>() }; capacity];
    let mut written = 0usize;
    let code = unsafe { ffi::rns_interfaces(node.handle(), raw.as_mut_ptr(), raw.len(), &mut written) };
    if code != 0 && code != 8 {
        map_code(code)?;
    }
    let mut out = Vec::with_capacity(written);
    for entry in raw.into_iter().take(written) {
        out.push(InterfaceInfo {
            name: cstr_field(&entry.name),
            type_name: cstr_field(&entry.type_name),
            online: entry.online != 0,
            enabled: entry.enabled != 0,
            rx_bytes: entry.rx_bytes,
            tx_bytes: entry.tx_bytes,
            rx_packets: entry.rx_packets,
            tx_packets: entry.tx_packets,
        });
    }
    Ok(out)
}
